crates/miden-objects/src/address/routing_parameters.rs

```diff
@@ -0,0 +1,301 @@
+use alloc::borrow::ToOwned;
+use alloc::string::{String, ToString};
+use alloc::vec::Vec;
+
+use bech32::primitives::decode::CheckedHrpstring;
+use bech32::{Bech32m, Hrp};
+
+use crate::AddressError;
+use crate::address::AddressInterface;
+use crate::errors::Bech32Error;
+use crate::note::NoteTag;
+use crate::utils::serde::{
+    ByteReader,
+    ByteWriter,
+    Deserializable,
+    DeserializationError,
+    Serializable,
+};
+use crate::utils::sync::LazyLock;
+
+/// The HRP used for encoding routing parameters.
+///
+/// `mrp` stands for Miden Routing Parameters.
+static ROUTING_PARAMETERS_HRP: LazyLock<Hrp> =
+    LazyLock::new(|| Hrp::parse("mrp").expect("hrp should be valid"));
+
+/// The separator character used in bech32.
+const BECH32_SEPARATOR: &str = "1";
+
+/// The value to encode the absence of a note tag routing parameter (i.e. `None`).
+///
+/// Note tag length is ensured to be <= [`NoteTag::MAX_LOCAL_TAG_LENGTH`] and so 1 << 5 = 32 is used
+/// to encode `None`.
+const ABSENT_NOTE_TAG_LEN: u8 = 1 << 5;
+
+/// The routing parameter key for the receiver profile.
+const RECEIVER_PROFILE_KEY: u8 = 0;
+
+/// TODO: Docs.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct RoutingParameters {
+    interface: AddressInterface,
+    note_tag_len: Option<u8>,
+}
```

Comment:
Why do we want `note_tag_len` to be optional? I'm thinking that if we go through the trouble of specifying routing parameters, we might as well specify note tag length. Are there any downsides to making both fields non-optional?

Also, should we not implement `Default` for this type? In my mind, not specifying routing parameters is equivalent to specifying some "universal" defaults. For example, when creating a note addressed to this address, we either get the tag length from the address or we have to assume it is some kind of a default value. Implementing `Default` would make this explicit, I think.

This will require bringing back `AddressInterface::Unspecified` - but I think this may be OK.